### ISSUE TITLE
How to modify gradients of an FSDP model?

### ISSUE THREAD

**Issue by eldarkurtic** [pytorch/pytorch#103958]
### 📚 The doc issue

I've initially posted the question on [forum](https://discuss.pytorch.org/t/modify-gradients-of-an-fsdp-model/182159) 7 days ago, but crossposting here as well for better visibility since I couldn't get any answers there. 

Hi everyone,
I have an FSDP model which has zeros in some of the `torch.nn.Linear.weight` parameters. During the training I would like to keep those parameters fixed to zeros, and to zero-out their gradients during backward as well. The specific use-case is: I am loading a pruned model and I want to fine-tune it with FSDP while keeping the pruning mask fixed. 

To achieve this I need to do two things: 
1) multiply parameters with the mask before the forward pass (so that all pruned weights remain pruned), 
2) multiply gradients of pruned parameters after the backward pass (so that gradients of pruned weights are zeros)
 
In the standard DDP training I would achieve this by:
1) registering forward pre-hook on `torch.nn.Linear` modules and multiplying weights with the mask before each forward pass,
2) registering a hook on the parameter `torch.nn.Linear.weight` and multiplying its gradient with the mask. 

For example:
```python
def keep_param_pruned(mask, module, input):
    with torch.no_grad():
        module.weight.data.mul_(mask.to(module.weight.device))

def keep_grad_pruned(mask, grad):
    return grad.mul_(mask.to(grad.device))

for n, m in model.named_modules():
    if isinstance(m, torch.nn.Linear):
        mask = m.weight > threshold
        m.register_forward_pre_hook(partial(keep_param_pruned, mask))
        m.weight.register_hook(partial(keep_grad_pruned, mask))
```

However, I am struggling to modify this idea to work with FSDP.  Any suggestions/ideas on what I am doing wrong or if there is a simpler way to achieve this without playing with hooks?

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by eldarkurtic** [comment]
Hi @awgu , a gentle reminder about this

**Comment by matanhs** [comment]
I have a similar issue trying to use gradient hooks with FSDP.
I attached the gradient hooks after applying the FSDP wrapper, however, the hooks never executed.
One workaround would be to extend the optimizer step function so it will modify the gradients in place. However, this will only work with gradient accumulation = 1 or require additional memory to store gradient accumulators.
Notably this functionality is also broken when using Deepspeed

Any word on fixing this issue?

**Comment by awgu** [comment]
Hi all. We hear the feature request, and we will look into it when we can.

**Comment by mori360** [comment]
We are committed to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
If you encounter any similar issues with FSDP2, please open a new issue with updated information.
